feat(learn): Add/Remove Tests for Imperial-Metric Converter Project (#39615)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/06-quality-assurance/quality-assurance-projects/metric-imperial-converter.english.md

```yml
tests:
  - text: 'I can GET /api/convert with a single parameter containing an accepted number and unit and have it converted. (Hint: Split the input by looking for the index of the first character which will mark the start of the unit)'
    testString: ''

>   - text: I can convert <code>'gal'</code> to <code>'L'</code> and vice versa. (1 gal to 3.78541 L)

Reviewer: Can we add the validation for the example URL? Like on the exercise tracker, to ensure campers cannot submit the example project and pass. Something like this:

```yaml
  - text: I can provide my own project, not the example URL.
    testString: "getUserInput => {
      assert(!(new RegExp('.*/metric-imperial-converter--freecodecamp\\.repl\\.co')).test(getUserInput('url')));
    }
    "
```

Author: Yeah, I can do that.

Author: Done.  New commit added.

        assert.equal(data4.returnNum, 6.21373); 
        assert.equal(data4.returnUnit, 'mi');
      } catch(xhr) {
        throw new Error(xhr.responseText || xhr.message);
      }
    }
    "
  - text: If my unit of measurement is invalid, returned will be <code>'invalid unit'</code>.
    testString: "async getUserInput => { 
      try {
        const data = await $.get(getUserInput('url') + '/api/convert?input=1min');
        assert(data.error === 'invalid unit' || data === 'invalid unit');
      } catch(xhr) {
        throw new Error(xhr.responseText || xhr.message);
      }
    }
    "

>   - text: If my number is invalid, returned with will 'invalid number'.

Reviewer 2: ```suggestion
  - text: If my number is invalid, returned will be <code>'invalid number'</code>.
```

    testString: "async getUserInput => { 
      try {
        const data = await $.get(getUserInput('url') + '/api/convert?input=1//2gal');
        assert(data.error === 'invalid number' || data === 'invalid number');
      } catch(xhr) {
        throw new Error(xhr.responseText || xhr.message);
      }
    }"

>   - text: If both are invalid, return will be 'invalid number and unit'.

Reviewer 2: ```suggestion
  - text: If both are invalid, return will be <code>'invalid number and unit'</code>.
```

    testString: "async getUserInput => {
      try {
        const data1 = await $.get(getUserInput('url') + '/api/convert?input=1/2mi');
        assert.approximately(data1.initNum, 1/2, 0.001);
        assert.approximately(data1.returnNum, 0.80467, 0.001);
        assert.equal(data1.returnUnit, 'km');
        const data2 = await $.get(getUserInput('url') + '/api/convert?input=1/5mi');
        assert.approximately(data2.initNum, 1/5, 0.1);
        assert.approximately(data2.returnNum, 0.32187, 0.001);
        assert.equal(data2.returnUnit, 'km');
        const data3 = await $.get(getUserInput('url') + '/api/convert?input=1.5/7km');
        assert.approximately(data3.initNum, 1.5/7, 0.001);
        assert.approximately(data3.returnNum, 0.13315, 0.001);
        assert.equal(data3.returnUnit, 'mi');
        const data4 = await $.get(getUserInput('url') + '/api/convert?input=3/2.7km');
        assert.approximately(data4.initNum, 3/2.7, 0.001);
        assert.approximately(data4.returnNum, 0.69041, 0.001);
        assert.equal(data4.returnUnit, 'mi');
      } catch(err) {
        throw new Error(err.responseText || err.message);
      }
    }
    "

>   - text: My return will consist of the initNum, initUnit, returnNum, returnUnit, and string spelling out units in format '{initNum} {initial_Units} converts to {returnNum} {return_Units}' with the result rounded to 5 decimals in the string.

Reviewer 2: ```suggestion
  - text: My return will consist of the <code>initNum</code>, <code>initUnit</code>, <code>returnNum</code>, <code>returnUnit</code>, and <code>string</code> spelling out units in format <code>'{initNum} {initial_Units} converts to {returnNum} {return_Units}'</code> with the result rounded to 5 decimals.
```

        assert.equal(data4.returnNum, 6.21373); 
        assert.equal(data4.returnUnit, 'mi');
      } catch(xhr) {
        throw new Error(xhr.responseText || xhr.message);
      }
    }
    "

>   - text: All incoming units should be accepted in both upper and lower case, but should be returned in both the <code>initUnit</code> and <code>returnUnit</code> in lower case, except for liter, which should be represented as an uppercase "L".

Reviewer: ```suggestion
  - text: All incoming units should be accepted in both upper and lower case, but should be returned in both the <code>initUnit</code> and <code>returnUnit</code> in lower case, except for liter, which should be represented as an uppercase <code>'L'</code>.
```

I would wrap this in code tags like the other mentions of units.

        assert.equal(data3.initUnit, 'L');
        assert.equal(data3.returnUnit, 'gal');
        const data4 = await $.get(getUserInput('url') + '/api/convert?input=10KM');
        assert.equal(data4.initUnit, 'km');
        assert.equal(data4.returnUnit, 'mi');
      } catch(xhr) {
        throw new Error(xhr.responseText || xhr.message);
      }
    }"
  - text: If my unit of measurement is invalid, returned will be <code>'invalid unit'</code>.
    testString: "async getUserInput => { 
      try {
        const data = await $.get(getUserInput('url') + '/api/convert?input=1min');
        assert(data.error === 'invalid unit' || data === 'invalid unit');
      } catch(xhr) {
        throw new Error(xhr.responseText || xhr.message);
      }
    }
    "
  - text: If my number is invalid, returned will be <code>'invalid number'</code>.
    testString: "async getUserInput => { 
      try {
        const data = await $.get(getUserInput('url') + '/api/convert?input=1//2gal');
        assert(data.error === 'invalid number' || data === 'invalid number');
      } catch(xhr) {
        throw new Error(xhr.responseText || xhr.message);
      }
    }"

>   - text: If both are invalid, return will be <code>'invalid number and unit'</code>.

Reviewer: ```suggestion
  - text: If both the unit and number are invalid, returned will be <code>'invalid number and unit'</code>.
```

Verbiage tweak here.

        assert.approximately(data3.initNum, 1.5/7, 0.001);
        assert.approximately(data3.returnNum, 0.13315, 0.001);
        assert.equal(data3.returnUnit, 'mi');
        const data4 = await $.get(getUserInput('url') + '/api/convert?input=3/2.7km');
        assert.approximately(data4.initNum, 3/2.7, 0.001);
        assert.approximately(data4.returnNum, 0.69041, 0.001);
        assert.equal(data4.returnUnit, 'mi');
      } catch(err) {
        throw new Error(err.responseText || err.message);
      }
    }
    "
  - text: My return will consist of the <code>initNum</code>, <code>initUnit</code>, <code>returnNum</code>, <code>returnUnit</code>, and <code>string</code> spelling out units in format <code>'{initNum} {initial_Units} converts to {returnNum} {return_Units}'</code> with the result rounded to 5 decimals.
    testString: "async getUserInput => { 
      try {
        const data = await $.get(getUserInput('url') + '/api/convert?input=2mi');
        assert.equal(data.initNum, 2);
        assert.equal(data.initUnit, 'mi');
        assert.approximately(data.returnNum, 3.21868, 0.001);
        assert.equal(data.returnUnit, 'km', 'returnUnit did not match');

>         assert.equal(data.string, '2 miles converts to 3.21868 kilometers')

Reviewer: Just a side thought here - in the interest of internationalisation, do we want to explicitly require the American spelling here? Other English dialects will potentially use `kilometres` instead.

Reviewer 3: Always use the American spelling.
